[Support inference params-5]Add docs for inference params (#8977)
mlflow/mlflow: docs/source/models.rst

---------------------------------
When working with ML models you often need to know some basic functional properties of the model
at hand, such as "What inputs does it expect?" and "What output does it produce?". MLflow models can
include the following additional metadata about model inputs, outputs and params that can be used by
downstream tooling:

* :ref:`Model Inference Params <inference-params>` - description of params used for model inference.
* :ref:`Model Signature <model-signature>` - description of a model's inputs, outputs and parameters.
* :ref:`Model Input Example <input-example>` - example of a valid model input.

.. _inference-params:

Model Inference Params
^^^^^^^^^^^^^^^^^^^^^^
Inference params are parameters that are passed to the model at inference time. These parameters
do not need to be specified when training the model, but could be useful for inference. With the 
advances in foundational models, more often "inference configuration" is used to modify the behavior 
of a model. In some cases, especially popular LLMs, the same model may require different parameter 

> configurations for different samples. 

Reviewer: ```suggestion
configurations for different samples at inference time. 
```

> With this newly introduced feature, you can now specify a dictionary of inference params during 
> model inference, and this could be extremely useful for LLMs. By passing different params such

Reviewer: ```suggestion
With this newly introduced feature, you can now specify a dictionary of inference params during 
model inference, providing a broader utility and improved control over the generated inference results, particularly for LLM use cases. By passing different params such
```

as ``temperature``, ``max_length``, etc. to the model at inference time, you can easily control 
the output of the model.

In order to use params at inference time, a valid :ref:`Model Signature <model-signature>` with 

> ``params`` must be defined. The params are passed to the model at inference time as a dictionary,

Reviewer: ```suggestion
``params`` must be defined. The params are passed to the model at inference time as a dictionary
```

and each param value will be validated against the corresponding param type defined in the model

> signature. Valid param types are ``Union[DataType, List[DataType], None]`` where ``DataType`` is

Reviewer: Instead of the typing signature here, could we just rely on the great bullet list that you have below and mention that it takes in either nothing (``None``), a ``DataType`` that is listed below, or a list of ``DataType``s that are listed below?

:py:class:`MLflow data types <mlflow.types.DataType>`. Please note that ``bytes`` are not supported.

A full list of supported param types is listed below:

* DataType.string or an array of DataType.string
* DataType.integer or an array of DataType.integer
* DataType.boolean or an array of DataType.boolean
* DataType.double or an array of DataType.double
* DataType.float or an array of DataType.float
* DataType.long or an array of DataType.long
* DataType.datetime or an array of DataType.datetime

.. note::
    When validating param values, the values will be converted to python native types.
    For example, ``np.float32(0.1)`` will be converted to ``float(0.1)``.

> 

Reviewer: Can we add a VERY simple example of logging a model with params and then calling the pyfunc representation of that model here just so that users have something visual to follow along with about what this feature is (it can be a unit-test friendly sort of example that uses a custom pyfunc just to illustrate the API behavior)?

---------------------------------
When working with ML models you often need to know some basic functional properties of the model
at hand, such as "What inputs does it expect?" and "What output does it produce?". MLflow models can
include the following additional metadata about model inputs, outputs and params that can be used by
downstream tooling:

* :ref:`Model Inference Params <inference-params>` - description of params used for model inference.
* :ref:`Model Signature <model-signature>` - description of a model's inputs, outputs and parameters.
* :ref:`Model Input Example <input-example>` - example of a valid model input.

.. _inference-params:

Model Inference Params
^^^^^^^^^^^^^^^^^^^^^^
Inference params are parameters that are passed to the model at inference time. These parameters
do not need to be specified when training the model, but could be useful for inference. With the 

> advances in foundational models, more often “inference configuration” is used to modify the behavior 

Reviewer 2: ```suggestion
advances in foundational models, more often "inference configuration" is used to modify the behavior 
```

of a model. In some cases, especially popular LLMs, the same model may require different parameter 
configurations for different samples. 

With this newly introduced feature, you can now specify a dictionary of inference params during 
model inference, and this could be extremely useful for LLMs. By passing different params such
as ``temperature``, ``max_length``, ``min_length``, ``num_beams``, ``repetition_penalty``, etc. to
the model at inference time, you can easily control the output of the model.

In order to use params at inference time, a valid :ref:`Model Signature <model-signature>` with 
``params`` must be defined. The params are passed to the model at inference time as a dictionary,
and each param value will be validated against the corresponding param type defined in the model

> signature. Valid param types are ``Union[DataType, List[DataType], None]`` where DataType is

Reviewer 2: ```suggestion
signature. Valid param types are ``Union[DataType, List[DataType], None]`` where ``DataType`` is
```

:py:class:`MLflow data types <mlflow.types.DataType>`. Please note that ``bytes`` are not supported.

> A full list of supported param types is listed below in the form of 
> :py:class:`ParamSpec <mlflow.types.ParamSpec>`, default values could be any valid value of the 
> specified type:
> 
> * ``ParamSpec("str_param", DataType.string, "DEFAULT_STRING", None)``
> * ``ParamSpec("int_param", DataType.integer, 1, None)``
> * ``ParamSpec("bool_param", DataType.boolean, True, None)``
> * ``ParamSpec("double_param", DataType.double, 1.0, None)``
> * ``ParamSpec("float_param", DataType.float, np.float32(0.1), None)``
> * ``ParamSpec("long_param", DataType.long, 100, None)``
> * ``ParamSpec("datetime_param", DataType.datetime, np.datetime64("2023-06-26 00:00:00"), None)``
> * ``ParamSpec("str_array", DataType.string, ["a", "b", "c"], (-1,))``
> * ``ParamSpec("int_array", DataType.integer, [1, 2, 3], (-1,))``
> * ``ParamSpec("bool_array", DataType.boolean, [True, False, True], (-1,))``
> * ``ParamSpec("double_array", DataType.double, [1.0, 2.0, 3.0], (-1,))``
> * ``ParamSpec("float_array", DataType.float, [1.0, 2.0, 3.0], (-1,))``
> * ``ParamSpec("long_array", DataType.long, [1, 2, 3], (-1,))``
> * ``ParamSpec("datetime_array", DataType.datetime, [np.datetime64("2023-06-26 00:00:00")], (-1,))``

Reviewer 2: Can I just list types?

```suggestion
A full list of supported param types is listed below in the form of 
:py:class:`ParamSpec <mlflow.types.ParamSpec>`, default values could be any valid value of the 
specified type:

- DataType.string
- DataType.integer
- ...
- An array of DataType.string
- An array of DataType.datetime
```

        (cm)", "type": "double"}, {"name": "petal length (cm)", "type": "double"}, {"name":
        "petal width (cm)", "type": "double"}, {"name": "class", "type": "string", "optional": "true"}]'
      outputs: '[{"type": "integer"}]'

>       params: None

Reviewer 2: ```suggestion
      params: null
```

  signature:
      inputs: '[{"name": "images", "dtype": "uint8", "shape": [-1, 28, 28, 1]}]'
      outputs: '[{"shape": [-1, 10], "dtype": "float32"}]'
      params: None

Signature with params Example
"""""""""""""""""""""""""""""
The params field is optional and is used to specify parameters that can be used for model inference.
Params accept scalar values of type :py:class:`MLflow data types <mlflow.types.DataType>`, or a list
of such values. The default value of a parameter is specified by setting the ``default`` field, and the value
should be of the type specified by ``type`` field. The ``shape`` field can be used to specify the shape 
of the value, it should be ``None`` for scalar values and ``(-1,)`` for a list.

.. code-block:: yaml

    signature:
        inputs: '[{"name": "text", "type": "string"}]'
        outputs: '[{"name": "output", "type": "string"}]'

>         params: '[{"name": "temperature", "type": "float", "default": 0.5, "shape": None},
>                   {"name": "top_k", "type": "integer", "default": 1, "shape": None},

Reviewer 2: ```suggestion
        params: '[{"name": "temperature", "type": "float", "default": 0.5, "shape": null},
                  {"name": "top_k", "type": "integer", "default": 1, "shape": null},
```

>                   {"name": "suppress_tokens", "type": "integer", "default": [101, 102], "shape": (-1,)}]'

Reviewer 2: ```suggestion
                  {"name": "suppress_tokens", "type": "integer", "default": [101, 102], "shape": [-1]}]'
```

`(-1,)` is not JSON-deserializable.

Author: Good catch!

For models with tensor-based signatures, type checking is strict (i.e an exception will be thrown if
the input type does not match the type specified by the schema).

Params Type and Shape Enforcement
"""""""""""""""""""""""""""""""""
The params types and shapes are checked against the signature.

MLflow verifies the compatibility of each parameter provided during inference by comparing its type and shape 
with those specified in the signature. Scalar values should have a shape of ``None``, while list values should have 
a shape of ``(-1,)``. If the parameter's type or shape is incompatible, an exception will be raised. 
Additionally, the value of the parameter is validated against the specified type in the signature. We attempt 
to convert the value to the specified type, and if this conversion fails, an MlflowException will be raised.

> 

Reviewer: Can we provide a list with generic examples of what parameter types are supported?
Can there also be an explanation added of why this feature exists and how it can be used as an intro to a new section above this? This is a pretty important and long asked-for feature that I think another `"""""` section is warranted to explain that inference parameters are supported now :)

    with mlflow.start_run():
        mlflow.tensorflow.log_model(model, "mnist_cnn", signature=signature)

Signature with params Example
"""""""""""""""""""""""""""""
The following example demonstrates how to store a model signature with params
for a simple transformers model:

.. code-block:: python

    import mlflow
    from mlflow.models import infer_signature
    import transformers

>     architecture = "mrm8488/t5-base-finetuned-common_gen"

Reviewer 2: How large is this model?

Author: roughly 1GB

    model = transformers.pipeline(
        task="text2text-generation",
        tokenizer=transformers.AutoTokenizer.from_pretrained(architecture),
        model=transformers.AutoModelWithLMHead.from_pretrained(architecture),

>     )

Reviewer 2: I got this error when I ran this example:

```
Traceback (most recent call last):
  File "a.py", line 7, in <module>
    tokenizer=transformers.AutoTokenizer.from_pretrained(architecture),
  File "/usr/local/lib/python3.8/site-packages/transformers/models/auto/tokenization_auto.py", line 637, in from_pretrained
    return tokenizer_class_fast.from_pretrained(pretrained_model_name_or_path, *inputs, **kwargs)
  File "/usr/local/lib/python3.8/site-packages/transformers/tokenization_utils_base.py", line 1777, in from_pretrained
    return cls._from_pretrained(
  File "/usr/local/lib/python3.8/site-packages/transformers/tokenization_utils_base.py", line 1932, in _from_pretrained
    tokenizer = cls(*init_inputs, **init_kwargs)
  File "/usr/local/lib/python3.8/site-packages/transformers/models/t5/tokenization_t5_fast.py", line 134, in __init__
    super().__init__(
  File "/usr/local/lib/python3.8/site-packages/transformers/tokenization_utils_fast.py", line 120, in __init__
    raise ValueError(
ValueError: Couldn't instantiate the backend tokenizer from one of: 
(1) a `tokenizers` library serialization file, 
(2) a slow tokenizer instance to convert or 
(3) an equivalent slow tokenizer class to instantiate and convert. 
You need to have sentencepiece installed to convert a slow tokenizer to a fast one.
```

        "num_beams": 5,
        "max_length": 30,
        "temperature": 0.62,
        "top_p": 0.85,
        "repetition_penalty": 1.15,
    }

    # infer signature with params
    signature = infer_signature(
        data,
        mlflow.transformers.generate_signature_output(model, data),
        params,
    )

    # save model with signature
    mlflow.transformers.save_model(
        model,
        "text2text",
        signature=signature,
    )
    pyfunc_loaded = mlflow.pyfunc.load_model("text2text")

    # predict with params
    result = pyfunc_loaded.predict(data, params=params)

The same signature can be created explicitly as follows:

.. code-block:: python

    from mlflow.models import ModelSignature
    from mlflow.types.schema import ColSpec, ParamSchema, ParamSpec, Schema

    input_schema = Schema([ColSpec(type="string")])
    output_schema = Schema([ColSpec(type="string")])
    params_schema = ParamSchema(
        [
            ParamSpec("top_k", "long", 2),
            ParamSpec("num_beams", "long", 5),
            ParamSpec("max_length", "long", 30),
            ParamSpec("temperature", "double", 0.62),
            ParamSpec("top_p", "double", 0.85),

>             ParamSpec("repetition_penalty", "double", 1.15),

Reviewer 2: Can we replace a few parameters with array-type params?

Once loaded, you can score the model by calling the :py:func:`predict <mlflow.pyfunc.PyFuncModel.predict>`
method, which has the following signature::

>   predict(data: Union[pandas.(Series | DataFrame), numpy.ndarray, csc_matrix, csr_matrix, List[Any], Dict[str, Any], str], params: Optional[Dict[str, Any]] = None) → Union[pandas.(Series | DataFrame), numpy.ndarray, list, str]

Reviewer 2: Can we split this into multiple lines?

<img width="1850" alt="image" src="https://github.com/mlflow/mlflow/assets/17039389/4db1e998-2951-4ab3-9d35-eced52dee3b7">

""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""
For `transformers` inference, there are two ways to pass in additional arguments to the pipeline.

* Use ``inference_config`` when saving/logging the model
* Specify params at inference time when calling ``predict()``

.. note::
    Model signature params passed in to ``predict`` function will override the values in inference_config.

* Using ``inference_config``

.. code-block:: python

    from mlflow.models import infer_signature
    from mlflow.transformers import generate_signature_output

    # Infer the signature
    signature = infer_signature(
        data,
        generate_signature_output(model, data),
    )

    # Define an inference_config
    inference_config = {
        "num_beams": 5,
        "max_length": 30,
        "do_sample": True,
        "remove_invalid_values": True,
    }

    # Saving inference_config with the model
    mlflow.transformers.save_model(
        model,
        path=model_path,
        inference_config=inference_config,
        signature=signature,
    )

    pyfunc_loaded = mlflow.pyfunc.load_model(model_path)
    # inference_config will be applied

>     result = pyfunc_loaded.predict(data)

Reviewer 2: I think a long complete example is better than a short incomplete example.

>     result = pyfunc_loaded.predict(data)

Author: Updated!

    inference_config = {
        "num_beams": 5,
        "max_length": 30,
        "do_sample": True,
        "remove_invalid_values": True,
    }

    # Saving inference_config with the model
    mlflow.transformers.save_model(
        model,
        path="text2text",
        inference_config=inference_config,
        signature=signature,
    )

    pyfunc_loaded = mlflow.pyfunc.load_model("text2text")
    # inference_config will be applied
    result = pyfunc_loaded.predict(data)

* Specifying params at inference time

.. code-block:: python

    import mlflow
    from mlflow.models import infer_signature
    from mlflow.transformers import generate_signature_output
    import transformers

    architecture = "mrm8488/t5-base-finetuned-common_gen"
    model = transformers.pipeline(
        task="text2text-generation",
        tokenizer=transformers.T5TokenizerFast.from_pretrained(architecture),
        model=transformers.T5ForConditionalGeneration.from_pretrained(architecture),
    )
    data = "pencil draw paper"

    # Infer the signature including params
    signature_with_params = infer_signature(
        data,
        generate_signature_output(model, data),

>         params=inference_config,

Reviewer: Does this run? Isn't inference_config defined on line 2911?

    )

    # Define an inference_config
    inference_config = {
        "num_beams": 5,
        "max_length": 30,
        "do_sample": True,
        "remove_invalid_values": True,
    }

    # Saving model without inference_config
    mlflow.transformers.save_model(
        model,
        path="text2text",
        signature=signature_with_params,
    )

    pyfunc_loaded = mlflow.pyfunc.load_model("text2text")
    # Passing inferenc_config dictionary as params to predict function directly

>     result = pyfunc_loaded.predict(data, params=inference_config)

Reviewer: Can we show just overriding a single parameter (or only 2 of the 4) and add a note that the logged default params will use the values that were specified during logging for the ``params`` values? I think that will help demonstrate to users that the concept of default values will use the pre-defined ones :)

\***** If using `pyfunc` in MLflow Model Serving for realtime inference, the raw audio in bytes format must be base64 encoded prior to submitting to the endpoint. String inputs will be interpreted as uri locations.

Using inference_config and model signature params for `transformers` inference
""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""
For `transformers` inference, there are two ways to pass in additional arguments to the pipeline.

> * Use inference_config when saving/logging the model

Reviewer 2: ```suggestion
* Use ``inference_config`` when saving/logging the model
```

> * Specify params at inference time when calling `predict()`

Reviewer 2: ```suggestion
* Specify params at inference time when calling ``predict()``
```

.. note::

>     Model signature params passed in to `predict` function will override the values in inference_config.

Reviewer 2: ```suggestion
    Model signature params passed in to ``predict`` function will override the values in inference_config.
```

> * Using inference_config

Reviewer 2: ```suggestion
* Using ``inference_config``
```
